ability to sort/merge by tag

pmarks opened the pull request:
Add a new -t TAG argument to sort and merge, allowing the file to be sorted by tag. For highly multiplexed BAM files (e.g. single-cell RNA-seq, or linked-reads), this is useful to group all the data for one 'barcode' together in the BAM file.

This emerged as a useful feature in discussion w/ @gavinha and @imielinski among others.

Any suggestions on how to improve the patch or add unit tests would be helpful.

walaj commented:
I second this PR. Very useful functionality, and it works in my tests (tried separately sorting by a ``Z`` tag and ``i`` tag).

walaj commented:
One small hangup is that the SAM format defines the following sort orders (``SO``): unknown (default), unsorted, queryname and coordinate.

The tag-sorted BAMs of course don't follow any of these. The format would have to be updated to specify tag-sorted info in the BAM header.

In the meantime, this PR should be amended to also change the sort order to ``unknown`` in the header. Right now, it stays as ``coordinate`` even after the tag sorting, which will confuse some tools once it finds alignments out of genomic order.

imielinski commented:
this is great!


Will give it a spin ..


Marcin

________________________________
From: Jeremiah Wala <notifications@github.com>
Sent: Saturday, May 6, 2017 12:06:40 PM
To: samtools/samtools
Cc: Marcin Imielinski; Mention
Subject: Re: [samtools/samtools] ability to sort/merge by tag (#675)


One small hangup is that the SAM format defines the following sort orders (SO): unknown (default), unsorted, queryname and coordinate.

The tag-sorted BAMs of course don't follow any of these. The format would have to be updated to specify tag-sorted info in the BAM header.

In the meantime, this PR should be amended to also change the sort order to unknown in the header. Right now, it stays as coordinate even after the tag sorting, which will confuse some tools once it finds alignments out of genomic order.

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/samtools/samtools/pull/675#issuecomment-299649425>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEtC4_9BWPrw8fHo8EmfqL3zBbP9aPZQks5r3JqQgaJpZM4NQ9xy>.

This electronic message is intended for the use of the named recipient only, and may contain information that is confidential, privileged or protected from disclosure under applicable law. If you are not the intended recipient, or an employee or agent responsible for delivering this message to the intended recipient, you are hereby notified that any reading, disclosure, dissemination, distribution, copying or use of the contents of this message including any of its attachments is strictly prohibited. If you have received this message in error or are not the named recipient, please notify us immediately by contacting the sender at the electronic mail address noted above, and destroy all copies of this message. Please note, the recipient should check this email and any attachments for the presence of viruses. The organization accepts no liability for any damage caused by any virus transmitted by this email.

pmarks commented:
@jkbonfield @jmarshall @jenniferliddle  - any comments on this PR?  Is there any additional process I need to follow to get this considered?

daviesrob commented on bam_sort.c:
```diff
@@ -1611,13 +1633,71 @@ static int change_SO(bam_hdr_t *h, const char *so)
+static inline int bam1_lt_by_tag(const bam1_p a, const bam1_p b) 
+{
+    uint8_t* aux_a = bam_aux_get(a, g_sort_tag);
+    uint8_t* aux_b = bam_aux_get(b, g_sort_tag);
```
bam_aux_get() does a linear scan through the read aux data, so putting two of them here is likely to slow the sort down a lot.  It would be better to replace the definition of bam1_p with something like this:
```
typedef struct bam1_p {
    const bam1_t *b;
    const uint8_t *tag;  // Or maybe a ptrdiff_t giving the offset of the tag in the aux data
} bam1_p;
```

The tag pointers can be populated when the bam records are read in, prior to the sort.  This will give a 2*log(n) reduction in the number of calls to bam_aux_get().

You would need to do the same thing during the merge stage, but note that merge can change the aux data in `bam_translate()`.  You would need to set the tag pointer **after** this has happened.

pmarks replied:
did a quick profile and indeed it's spending a ~60% of it's time in bam_aux_get(), so I will implement your suggestion of a wrapper struct that is pre-populated with the appropriate tag pointer.

daviesrob commented on bam_sort.c:
```diff
@@ -1611,13 +1633,71 @@ static int change_SO(bam_hdr_t *h, const char *so)
+    uint8_t* aux_a = bam_aux_get(a, g_sort_tag);
+    uint8_t* aux_b = bam_aux_get(b, g_sort_tag);
+
+    if (aux_a == 0 && aux_b > 0) {
```
Should be:
```
if (aux_a == NULL && aux_b != NULL) {
```

The comparison `aux_b > 0` is undefined behaviour, see [CERT ARR36-C](https://www.securecoding.cert.org/confluence/display/c/ARR36-C.+Do+not+subtract+or+compare+two+pointers+that+do+not+refer+to+the+same+array)

daviesrob replied:
Should be:
```
if (aux_a == NULL && aux_b != NULL) {
```

The comparison `aux_b > 0` is undefined behaviour, see [CERT ARR36-C](https://www.securecoding.cert.org/confluence/display/c/ARR36-C.+Do+not+subtract+or+compare+two+pointers+that+do+not+refer+to+the+same+array)

daviesrob commented on bam_sort.c:
```diff
@@ -1611,13 +1633,71 @@ static int change_SO(bam_hdr_t *h, const char *so)
+        return bam1_lt_core(a,b);
+    } 
+
+    if (*aux_a < *aux_b) {
```
This doesn't quite work, as for integer tags `*aux_a` and `*aux_b` can have different types (say c and i), but still be comparable.  You might want to cover the integer case before this (and do we want to allow comparison of floats and integers?)

pmarks replied:
Our use case is currently for string tags that encode a barcode sequence, but I can imagine changing to an integer barcode for efficiency. Not sure about floats, but seems worth handling all in one shot.

daviesrob commented on bam_sort.c:
```diff
@@ -1611,13 +1633,71 @@ static int change_SO(bam_hdr_t *h, const char *so)
+        return 0;
+    } else {
+        if (*aux_a == 'c' || *aux_a == 'C' || *aux_a == 's' || *aux_a == 'S' || *aux_a == 'i' || *aux_a == 'I') {
+            int32_t va = bam_aux2i(aux_a);
```
Should be `int64_t`.

daviesrob replied:
Should be `int64_t`.

daviesrob commented on bam_sort.c:
```diff
@@ -1756,6 +1837,12 @@ int bam_sort_core_ext(int is_by_qname, const char *fn, const char *prefix,
+    if (sort_by_tag) {
+        g_is_by_tag = 1;
+        g_sort_tag[0] = sort_by_tag[0];
+        g_sort_tag[1] = sort_by_tag[1];
```
`strncpy()` again.

pmarks commented:
@daviesrob: all comments addressed in f1885e0 - ready for another review. Thanks!

pmarks commented:
fixed a bug in merge, and added a merge test. (also resolved a conflict with a change to a test BAM from another commit to develop). ready for review.

pmarks commented:
@daviesrob ping

daviesrob commented:
Sorry, I've been a bit busy with some other stuff for a while.  But I haven't forgotten this, and hope to get it reviewed and in by the end of this week.

daviesrob commented:
This looks OK now, and the extra overhead for the tag sort doesn't look too bad.

I've added a few small updates:

- The tests were comparing to compressed bam files.  This can be a problem if a different version of zlib (for example [cloudflare's optimized one](https://github.com/cloudflare/zlib)) is used as the compressed data can be different even though the uncompressed data is the same.  I've changed them to use sam instead.  (As an aside, I see a few other expected bam files have sneaked in elsewhere.  They will also need to be changed.)

- I've allowed float tags to be compared to integer ones, as it's not too difficult.

- I've added documentation to the manual page, including a description of the sorting rules.

The revision history for this pull request is a bit messy, so I suspect I'll squash it on merging.  In general I think it's best if you don't try to merge changes from develop into pull requests as it keeps things cleaner.  We can easily sort out any issues at the point when we merge the whole thing.  If you really do want to be up to date, a better alternative is to rebase the whole thing and then force push it on top of the existing PR.

pmarks commented on bam_sort.c:
```diff
@@ -1664,7 +1664,7 @@ static inline int bam1_lt_core(const bam1_p a, const bam1_p b)
     if (*aux == 'c' || *aux == 'C' || *aux == 's' || *aux == 'S' || *aux == 'i' || *aux == 'I') {
         return 'c';
-    } else if (*aux == 'f' || *aux == 'F') {
+    } else if (*aux == 'f' || *aux == 'd') {
```
Should this be (*aux == 'f' || *aux == 'F' || *aux == 'd') ?

daviesrob replied:
I don't think so.  [The sam specification](http://samtools.github.io/hts-specs/SAMv1.pdf) lists `AcCsSiIfZHB` as valid type specifiers on page 12.  `d` is a historic samtools/htslib extension for "double" - I don't think it's ever written, but support for reading it is there.

pmarks commented:
@daviesrob great, thanks!  Noticed one thing in the aux type normalization - is 'F' a valid type that should be lumped in with 'f'?  Happy to squish on merge -- I'll be more careful w/ the history.  Anything else you need from me?

daviesrob commented:
We're pretty much there.

There is one other issue I noticed.  I'm not sure if it's right to relax the lost_coord_sort test in `bam_merge_core2()`.  When sorting by tag, if the secondary sort is by position it will still get messed up if the references are in a different order in the source files.

daviesrob commented:
Spotted a problem with merge by tag then name.  Should be sorted now.

daviesrob commented:
Merged as 3c7e138eccd60ac4bbceb1523eaabb0435fe0050.

Thanks for the pull request, but in future please don't merge in develop.  It took quite a lot of git-fu to get it into a reasonable state...
